src/system/SystemPacketBuffer.cpp

```diff
@@ -541,24 +523,66 @@ PacketBufferHandle PacketBufferHandle::CloneData()
         // We do not clone an entire chain.
         return PacketBufferHandle();
     }
+    return NewWithData(mBuffer->Start(), mBuffer->DataLength());
+}
+
+#if CHIP_SYSTEM_CONFIG_USE_LWIP && LWIP_PBUF_FROM_CUSTOM_POOLS
+
+void PacketBufferHandle::RightSizeForLwIPCustomPools()
+{
+    PacketBuffer * lNewPacket = static_cast<PacketBuffer *>(pbuf_rightsize((struct pbuf *) mBuffer, -1));
+    if (lNewPacket != mBuffer)
+    {
+        mBuffer = lNewPacket;
+        SYSTEM_STATS_UPDATE_LWIP_PBUF_COUNTS();
+        ChipLogProgress(chipSystemLayer, "PacketBuffer: RightSize Copied");
+    }
+}
+
+#elif CHIP_SYSTEM_CONFIG_PACKETBUFFER_MAXALLOC == 0
 
-    PacketBufferHandle other = PacketBuffer::New();
-    if (other.IsNull())
+// Number of unused bytes below which \c RightSizeForMemoryAlloc() won't bother reallocating.
+constexpr uint16_t kRightSizingThreshold = 16;
+
+void PacketBufferHandle::RightSizeForMemoryAlloc()
+{
+    // Require a single buffer with no other references.
+    if ((mBuffer == nullptr) || (mBuffer->next != nullptr) || (mBuffer->ref != 1))
     {
-        return other;
+        return;
     }
 
-    if (other->AvailableDataLength() < mBuffer->DataLength())
+    // Reallocate only if enough space will be saved.
+    uint8_t * const start = reinterpret_cast<uint8_t *>(mBuffer) + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE;
+    uint8_t * const payload = reinterpret_cast<uint8_t *>(mBuffer->payload);
+    const uint16_t usedSize = static_cast<uint16_t>(payload - start + mBuffer->len);
+    if (usedSize + kRightSizingThreshold > mBuffer->alloc_size)
     {
-        other = nullptr;
-        return other;
+        return;
     }
 
-    memcpy(other->Start(), mBuffer->Start(), mBuffer->DataLength());
-    other->SetDataLength(mBuffer->DataLength());
+    const uint16_t blockSize = usedSize + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE;
+    PacketBuffer * newBuffer = reinterpret_cast<PacketBuffer *>(chip::Platform::MemoryAlloc(blockSize));
```

Comment:
Could this use a Realloc-like API (did not check if we have it) followed by a memmove below?